fix nvidia 340 kernel module compilation error

MortezaRamezani opened the pull request:
This PR is a followup to this #6384. However I wasn't able to load the legacy340 on my old machine yet.

dmanlfc commented:
@MortezaRamezani thanks looks good, why comment out these lines...
	
```
#mv -f $(TARGET_DIR)/lib/modules/$(KVER)/extra/nvidia-modeset.ko \
	#    $(TARGET_DIR)/usr/share/nvidia/modules/nvidia340-modeset-legacy.ko
	#mv -f $(TARGET_DIR)/lib/modules/$(KVER)/extra/nvidia-drm.ko \
	#    $(TARGET_DIR)/usr/share/nvidia/modules/nvidia340-drm-legacy.ko	
	#mv -f $(TARGET_DIR)/lib/modules/$(KVER)/extra/nvidia-uvm.ko \
	#    $(TARGET_DIR)/usr/share/nvidia/modules/nvidia340-uvm-legacy.ko
```

MortezaRamezani commented:
The provided libs and kernel modules are different from 340 and 390. I believe the code was originally adapted from 390. I tried to keep them as close as possible. But I think there are some name differences as well. I am working to see if I can fix it.
Here's the list of files for 340 and 390:

```
nvidia340:                          nvidia390:

conftest                            common
conftest.h                          conftest
conftest.sh                         conftest1692980.c
cpuopsys.h                          conftest.sh
dkms.conf                           dkms.conf
gcc-version-check.c                 Kbuild
g_nvreadme.h                        Makefile
Makefile                            modules.order
modules.order                       Module.symvers
Module.symvers                      nv_compiler.h
nv-acpi.c                           nvidia
nv-acpi.o                           nvidia-drm
nv.c                                nvidia-drm.ko
nv-chrdev.c                         nvidia-drm.mod
nv-chrdev.o                         nvidia-drm.mod.c
nv_compiler.h                       nvidia-drm.mod.o
nv-cray.c                           nvidia-drm.o
nv-cray.o                           nvidia.ko
nv-dma.c                            nvidia.mod
nv-dma.o                            nvidia.mod.c
nv-drm.c                            nvidia-modeset
nv-drm.o                            nvidia-modeset.ko
nverror.h                           nvidia-modeset.mod
nv-frontend.c                       nvidia-modeset.mod.c
nv-frontend.h                       nvidia-modeset.mod.o
nv-frontend.o                       nvidia-modeset.o
nv_gpu_ops.h                        nvidia.mod.o
nvgputypes.h                        nvidia.o
nv-gvi.c                            nvidia-uvm
nv-gvi.o                            nvidia-uvm.ko
nv.h                                nvidia-uvm.mod
nv-i2c.c                            nvidia-uvm.mod.c
nv-i2c.o                            nvidia-uvm.mod.o
nvidia.ko                           nvidia-uvm.o
nvidia.mod
nvidia.mod.c
nvidia.mod.o
nvidia-modules-common.mk
nvidia.o
nv-kernel.o
nv-linux.h
nv-memdbg.h
nv-mempool.c
nv-mempool.o
nv-misc.h
nv-mmap.c
nv-mmap.o
nv.o
nv-p2p.c
nv-p2p.h
nv-p2p.o
nv-pat.c
nv-pat.h
nv-pat.o
nv-procfs.c
nv-procfs.o
nv-proto.h
nv-reg.h
nv-time.h
nvtypes.h
nv-usermap.c
nv-usermap.o
nv_uvm_interface.c
nv_uvm_interface.h
nv_uvm_interface.o
nv-vm.c
nv-vm.o
nv-vtophys.c
nv-vtophys.o
os-interface.c
os-interface.h
os-interface.o
os-mlock.c
os-mlock.o
os-pci.c
os-pci.o
os-registry.c
os-registry.o
os-smp.c
os-smp.o
os-usermap.c
os-usermap.o
rmil.h
rmretval.h
uvm
xapi-sdk.h
```

```
nvidia340:                          nvidia390:

32                                  10_nvidia.json
glext.h                             10_nvidia_wayland.json
gl.h                                32
glxext.h                            glext.h
glx.h                               gl.h
html                                glxext.h
kernel                              glx.h
libcuda.so.340.108                  html
libEGL.so.340.108                   kernel
libGLESv1_CM.so.340.108             libcuda.so.390.157
libGLESv2.so.340.108                libEGL_nvidia.so.390.157
libGL.la                            libEGL.so.1.1.0
libGL.so.340.108                    libEGL.so.390.157
libglx.so.340.108                   libGLdispatch.so.0
libnvcuvid.so.340.108               libGLESv1_CM_nvidia.so.390.157
libnvidia-cfg.so.340.108            libGLESv1_CM.so.1.2.0
libnvidia-compiler.so.340.108       libGLESv2_nvidia.so.390.157
libnvidia-eglcore.so.340.108        libGLESv2.so.2.1.0
libnvidia-encode.so.340.108         libGL.la
libnvidia-fbc.so.340.108            libGL.so.1.7.0
libnvidia-glcore.so.340.108         libGL.so.390.157
libnvidia-glsi.so.340.108           libglvnd_install_checker
libnvidia-ifr.so.340.108            libGLX_nvidia.so.390.157
libnvidia-ml.so.340.108             libGLX.so.0
libnvidia-opencl.so.340.108         libglx.so.390.157
libnvidia-tls.so.340.108            libnvcuvid.so.390.157
libnvidia-wfb.so.340.108            libnvidia-cfg.so.390.157
libOpenCL.so.1.0.0                  libnvidia-compiler.so.390.157
libvdpau_nvidia.so.340.108          libnvidia-eglcore.so.390.157
libvdpau.so.340.108                 libnvidia-egl-wayland.so.1.0.2
libvdpau_trace.so.340.108           libnvidia-encode.so.390.157
LICENSE                             libnvidia-fatbinaryloader.so.390.15
makeself-help-script.sh             libnvidia-fbc.so.390.157
makeself.sh                         libnvidia-glcore.so.390.157
mkprecompiled                       libnvidia-glsi.so.390.157
monitoring.conf                     libnvidia-gtk2.so.390.157
nvidia-application-profiles-340.108 libnvidia-gtk3.so.390.157
nvidia-application-profiles-340.108 libnvidia-ifr.so.390.157
nvidia-bug-report.sh                libnvidia-ml.so.390.157
NVIDIA_Changelog                    libnvidia-opencl.so.390.157
nvidia-cuda-mps-control             libnvidia-ptxjitcompiler.so.390.157
nvidia-cuda-mps-control.1.gz        libnvidia-tls.so.390.157
nvidia-cuda-mps-server              libnvidia-wfb.so.390.157
nvidia-debugdump                    libOpenCL.so.1.0.0
nvidia_drv.so                       libOpenGL.so.0
nvidia.icd                          libvdpau_nvidia.so.390.157
nvidia-installer                    LICENSE
nvidia-installer.1.gz               makeself-help-script.sh
nvidia-modprobe                     makeself.sh
nvidia-modprobe.1.gz                mkprecompiled
nvidia-persistenced                 nvidia-application-profiles-390.157
nvidia-persistenced.1.gz            nvidia-application-profiles-390.157
nvidia-persistenced-init.tar.bz2    nvidia-bug-report.sh
nvidia-settings                     NVIDIA_Changelog
nvidia-settings.1.gz                nvidia-cuda-mps-control
nvidia-settings.desktop             nvidia-cuda-mps-control.1.gz
nvidia-settings.png                 nvidia-cuda-mps-server
nvidia-smi                          nvidia-debugdump
nvidia-smi.1.gz                     nvidia-drm-outputclass.conf
nvidia-xconfig                      nvidia_drv.so
nvidia-xconfig.1.gz                 nvidia.icd
pci.ids                             nvidia_icd.json.template
pkg-history.txt                     nvidia-installer
README.txt                          nvidia-installer.1.gz
tls                                 nvidia-modprobe
tls_test                            nvidia-modprobe.1.gz
tls_test_dso.so                     nvidia-persistenced
                                    nvidia-persistenced.1.gz
                                    nvidia-persistenced-init.tar.bz2
                                    nvidia-settings
                                    nvidia-settings.1.gz
                                    nvidia-settings.desktop
                                    nvidia-settings.png
                                    nvidia-smi
                                    nvidia-smi.1.gz
                                    nvidia-xconfig
                                    nvidia-xconfig.1.gz
                                    pkg-history.txt
                                    README.txt
                                    tls
                                    tls_test
                                    tls_test_dso.so
```

dmanlfc commented:
ok, the `batocera-nvidia` script needs adapting for the 340 too.
and `batocera-boot.conf` needs to commented option highlighted

MortezaRamezani commented:
I updated those files, fixed the problem for building `uvm` as well. Apparently 340 is using old kernel module build system and wasn't executing most of the code in Makefile. However I'm still not sure if my change is taking place, because when I set the driver to legacy340 in boot, I can't see it in log. I might need to do cleanbuild. For now mark it as wip.

dmanlfc commented:
all good, take your time to get it right.
we do it this way so the image can be portable across different hardware if needed & possibly different nvidia cards.
the batocera.log should give information if you let it autodetect based on the 340 list file.

dmanlfc commented:
```
	    if grep -q -i $PCIDEVID $LISTDIR/legacy340.list 2>/dev/null
	    then
            # check card name
            echo "Detected PCI device ID in legacy340 list" >> $LOG
            if grep -q -i $CARDNAME $LISTDIR/legacy.list 2>/dev/null
            then
                echo "Detected card name in legacy340 list" >> $LOG
                MODE=legacy340
                break
            else
                echo "Card name not detected in legacy340 list" >> $LOG
            fi
	    fi
```

MortezaRamezani commented:
Yeah, I need to rebuild `batocera-nvidia` to apply those change. Nevertheless, I was able to load the driver using `insmod` but it's not loading on startup and no luck with `modeprob`. I think at least the driver build was successful, but I'll look into the automatic loading later.

MortezaRamezani commented:
This should compile, install and enable `legacy340` fine now. The kernel module is also loaded and `nvidia-smi` is showing the correct driver.
```
[root@BATOCERA /userdata/system]# ./nvidia-smi
Thu Feb  9 04:19:07 2023
+------------------------------------------------------+
| NVIDIA-SMI 340.107    Driver Version: 340.108        |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GT 330M     Off  | 0000:01:00.0     N/A |                  N/A |
| N/A   50C    P0    N/A /  N/A |      2MiB /  1023MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
+-----------------------------------------------------------------------------+
```


However on my old vaio laptop it gets stuck at Nvidia logo, which was something that I faced before with proprietary driver, and I hope other people who need this driver won't face.

nadenislamarre commented:
@dmanlfc for me it is ok for 36. as you prefer.

dmanlfc commented:
@nadenislamarre looks good. I'm still away for close to a week.
please merge & have some users on discord - beta test it with compatible legacy nvidia devices.
